File: stack/visible_people.py

def count_visibilities(heights: list[int]) -> list[int]:  # LeetCode Q.1944.
    stack = []  # Height decreasing monotonic stack.
    visibilities = []
    for height in heights[::-1]:  # Backward iteration.
        visibilities.append(0)  # My (if I am the rightmost) min visibility is 0.

        while stack and stack[-1] <= height:  # People on my left side can't see them.
            stack.pop(-1)
            visibilities[-1] += 1

        if stack:  # I can always see my next taller person (if exists).
            visibilities[-1] += 1

        stack.append(height)

    return visibilities[::-1]  # Revert back to original order.

File: stack/visible_people/visible_people.cpp
#include <vector>
#include <stack>
using namespace std;

vector<int> count_visibilities(vector<int> &heights) // LeetCode Q.1944.
{
    stack<int> stack; // Heights decreasing monotonic stack.
    vector<int> visible_people;

    reverse(heights.begin(), heights.end());
    for (auto height : heights)
    {
        visible_people.push_back(0);
        while (!stack.empty() && stack.top() < height)
        {
            visible_people.back()++;
            stack.pop();
        }

        if (!stack.empty())
        {
            visible_people.back()++;
            if (stack.top() == height)
            {
                stack.pop();
            }
        }

        stack.push(height);
    }

    reverse(visible_people.begin(), visible_people.end());
    return visible_people;
}
